sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+    0x80, 0xcb, 0x7b, 0xde, 0xd7, 0x3f, 0x4d, 0x93, 0x4d, 0x59, 0x79, 0x88,
+    0x24, 0xe7, 0x68, 0x8b, 0x7a, 0x78, 0xb7, 0x07, 0x09, 0x26, 0xcf, 0x6b,
+    0x52, 0xd9, 0x4c, 0xd3, 0x33, 0xdf, 0x2e, 0x0d, 0x3b, 0xab, 0x45, 0x85,
+    0xc2, 0xc2, 0x19, 0xe5, 0xc7, 0x2b, 0xb0, 0xf6, 0xcb, 0x06, 0xf6, 0xe2,
+    0xf5, 0xb1, 0xab, 0xef, 0x6f, 0xd8, 0x23, 0xfd,
+};
+static size_t uart_tx_bytes_written;
+
+// A set of bytes actually received over RX.
+static uint8_t uart_rx_data[UART_DATASET_SIZE];
+
+// The set of bytes expected to be received over RX.
+static const uint8_t exp_uart_rx_data[UART_DATASET_SIZE] = {
+    0x1b, 0x95, 0xc5, 0xb5, 0x8a, 0xa4, 0xa8, 0x9f, 0x6a, 0x7d, 0x6b, 0x0c,
+    0xcd, 0xd5, 0xa6, 0x8f, 0x07, 0x3a, 0x9e, 0x82, 0xe6, 0xa2, 0x2b, 0xe0,
+    0x0c, 0x30, 0xe8, 0x5a, 0x05, 0x14, 0x79, 0x8a, 0xFf, 0x88, 0x29, 0xda,
+    0xc8, 0xdd, 0x82, 0xd5, 0x68, 0xa5, 0x9d, 0x5a, 0x48, 0x02, 0x7f, 0x24,
+    0x32, 0xaf, 0x9d, 0xca, 0xa7, 0x06, 0x0c, 0x96, 0x65, 0x18, 0xe4, 0x7f,
+    0x26, 0x44, 0xf3, 0x14, 0xC1, 0xe7, 0xd9, 0x82, 0xf7, 0x64, 0xe8, 0x68,
+    0xf9, 0x6c, 0xa9, 0xe7, 0xd1, 0x9b, 0xac, 0xe1, 0xFd, 0xd8, 0x59, 0xb7,
+    0x8e, 0xdc, 0x24, 0xb8, 0xa7, 0xaf, 0x20, 0xee, 0x6c, 0x61, 0x48, 0x41,
+    0xB4, 0x62, 0x3c, 0xcb, 0x2c, 0xbb, 0xe4, 0x44, 0x97, 0x8a, 0x5e, 0x2f,
+    0x7f, 0x2b, 0x10, 0xcc, 0x7d, 0x89, 0x32, 0xfd, 0xfd, 0x58, 0x7f, 0xd8,
+    0xc7, 0x33, 0xd1, 0x6a, 0xc7, 0xba, 0x78, 0x69};
+static size_t uart_rx_bytes_read;
+
+// Set our expectation & event indications of the interrupts we intend to
+// exercise in this test.
+static bool uart_irq_tx_watermark_fired;
+static bool uart_irq_rx_watermark_fired;
+static bool uart_irq_tx_empty_fired;
+static bool uart_irq_rx_overflow_fired;
+
+/**
+ * Send data out over TX without blocking.
+ *
+ * Non-blocking function that writes uart_tx_data bytes to be sent out of UART.
+ * It stops when the TX fifo is full. As and when the TX watermark interrupt
+ * fires, the test calls this function to resume sending the remaining bytes.
+ */
+void send_tx_data_nonblocking(void) {
+  size_t bytes_remaining = sizeof(uart_tx_data) - uart_tx_bytes_written;
+  // Return if we have no data to send.
+  if (bytes_remaining == 0)
+    return;
+  // Send the remaining uart_tx_data.
+  size_t bytes_written_isr = 0;
+  bool status =
+      dif_uart_bytes_send(&uart0, &uart_tx_data[uart_tx_bytes_written],
+                          bytes_remaining, &bytes_written_isr);
+  CHECK(status, "dif_uart_bytes_send failed");
+  uart_tx_bytes_written += bytes_written_isr;
+}
+
+/**
+ * Receive data in over RX without blocking.
+ *
+ * Non-blocking function that reads bytes received over UART RX port. It reads
+ * as many bytes as there are available in the RX FIFO and returns the number of
+ * bytes read, which is used the next time this function is called. Eventually,
+ * all UART_DATASET_SIZE bytes need to be read.
+ */
+void receive_rx_data_nonblocking(void) {
```

Comment:
In this routine and others, it would be good if the DIF owner (@mcy in this case) were to keep an eye if there are things the DV team is creating in local test routines that could/should be absorbed into the DIFs. That, or if the DV team is duplicating stuff that the DIF already contains. @gkelly for visibility.

Reply:
I don't own UART, actually. That's @silvestrst.